bittide/tests/Tests/DoubleBufferedRam.hs

```diff
@@ -885,3 +891,169 @@ wbStorageProtocolsModel = property $ do
+compareTransWithUndef :: (KnownNat addrW, KnownNat selWidth, KnownNat n) =>
+  (Transaction addrW selWidth (BitVector n)) -> (Transaction addrW selWidth (BitVector n)) -> Bool
+compareTransWithUndef transA transB =
+  case (transA, transB) of
+    ((WriteSuccess mA _), (WriteSuccess mB _)) ->
+        checkField "addr" addr mA mB &&
+        checkField "buSelect" busSelect mA mB &&
+        checkField "writeData" writeData mA mB
+    ((ReadSuccess mA sA), (ReadSuccess mB sB)) ->
+        checkField "addr" addr mA mB &&
+        checkField "busSelect" busSelect mA mB &&
+        checkBitVectorField readData sA sB
+    ((Error _), (Error _)) -> True
+    ((Retry _), (Retry _)) -> True
+    ((Stall _), (Stall _)) -> True
+    ((Illegal _ _), (Illegal _ _)) -> True
+    ((Ignored _), (Ignored _)) -> True
+    (_, _) -> False
+
+checkBitVectorField :: (KnownNat n) => (t -> BitVector n) -> t -> t -> Bool
+checkBitVectorField f a b =
+    (f a) `eqBitVec` (f b)
+
+-- | Behavioral test for 'wbStorageTDP', it checks whether the behavior of
+-- 'wbStorageTDP' matches the 'wbStorageTDPBehaviorModel'.
+wbStorageTDPBehavior :: Property
+wbStorageTDPBehavior = verifiedTermination . withTests 1000 . property $ do
+  nWordsA <- forAll $ Gen.enum 2 32
+  nWordsB <- forAll $ Gen.enum 2 32
+  case (TN.someNatVal (nWordsA - 2), TN.someNatVal (nWordsB - 2)) of
+    (SomeNat (addSNat d2 . snatProxy -> nWordsA0), SomeNat (addSNat d2 . snatProxy -> nWordsB0)) -> go nWordsA0 nWordsB0
+ where
+  go :: forall wordsA wordsB m . (KnownNat wordsA, 2 <= wordsA, KnownNat wordsB, 2 <= wordsB, Monad m) => SNat wordsA -> SNat wordsB -> PropertyT m ()
+  go SNat SNat = do
+    wbRequestsNum <- forAll $ Gen.integral (Range.linearFrom 0 (0) 32)
+    wbRequestsA <- forAll $ Gen.list (Range.singleton wbRequestsNum)
+      (genWishboneTransfer @32 (natToNum @wordsA) (genDefinedBitVector @32))
+    wbRequestsB <- forAll $ Gen.list (Range.singleton wbRequestsNum)
+      (genWishboneTransfer @32 (natToNum @wordsB) (genDefinedBitVector @32))
+
+    cover 5 "equal address" $ or (L.zipWith checkAddressEqual (L.map fst wbRequestsA) (L.map fst wbRequestsB))
+
+    F.for_ (L.map (getAddressWbRequest . fst) (wbRequestsA L.++ wbRequestsB)) $ \x -> do
+        classify "low address" $ x == minBound
+        classify "mid address" $ x > minBound && x < maxBound
+        classify "high address" $ x == maxBound
+
+    F.for_ (L.map (getByteEnableWbRequest . fst) (wbRequestsA L.++ wbRequestsB)) $ \x -> do
+        classify "ByteEnable: 0000" $ x == 0
+        classify "ByteEnable: 0001" $ x == 1
+        classify "ByteEnable: 0010" $ x == 2
+        classify "ByteEnable: 0100" $ x == 4
+        classify "ByteEnable: 1000" $ x == 8
+
+    let
+      masterA = driveStandard defExpectOptions wbRequestsA
+      masterB = driveStandard defExpectOptions wbRequestsB
+      master = prod2C masterA masterB
+      slave = wbStorageTDPC @System @System @32 clockGen resetGen enableGen clockGen resetGen enableGen
```

Comment:
Have you considered using `circuit-notation`? Example: https://github.com/bittide/bittide-hardware/blob/staging/bittide/tests/Tests/Wishbone.hs#L108-L123

Reply:
I have not. For now, it is consistent with the rest of the module, maybe refactor in the future?